catalog/core/catalog-core-directorymonitor/src/test/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserverTest.java

```diff
@@ -0,0 +1,1006 @@
+    observer.checkAndNotify();
+    init();
+
+    Stream.of(files).forEach(this::fileDelete);
+    timesToFail.set(3);
+    for (int i = 0; i < 4; i++) {
+      observer.checkAndNotify();
+    }
+
+    verify(fileListener, never()).onFileCreate(any(File.class), any(Synchronization.class));
+    verify(fileListener, never()).onFileChange(any(File.class), any(Synchronization.class));
+    verify(fileListener, times(files.length + failures))
+        .onFileDelete(any(File.class), any(Synchronization.class));
+  }
+
+  @Test
+  public void testCreateThreadSafety() throws Exception {
+
+    File[] files = initFiles(1, monitoredDirectory, "file00");
+
+    runThreads(observer::checkAndNotify, 10);
+
+    verify(fileListener, times(files.length))
+        .onFileCreate(any(File.class), any(Synchronization.class));
+    verify(fileListener, never()).onFileChange(any(File.class), any(Synchronization.class));
+    verify(fileListener, never()).onFileDelete(any(File.class), any(Synchronization.class));
+  }
+
+  @Test
+  public void testModifyThreadSafety() throws Exception {
+
+    File[] files = initFiles(1, monitoredDirectory, "file00");
+    observer.checkAndNotify();
+    init();
+
+    Stream.of(files).forEach(this::changeData);
+
+    runThreads(observer::checkAndNotify, 10);
+
+    verify(fileListener, never()).onFileCreate(any(File.class), any(Synchronization.class));
+    verify(fileListener, times(files.length))
+        .onFileChange(any(File.class), any(Synchronization.class));
+    verify(fileListener, never()).onFileDelete(any(File.class), any(Synchronization.class));
+  }
+
+  @Test
+  public void testDeleteThreadSafety() throws Exception {
+
+    File[] files = initFiles(1, monitoredDirectory, "file00");
+    observer.checkAndNotify();
+    init();
+
+    Stream.of(files).forEach(this::fileDelete);
+
+    runThreads(observer::checkAndNotify, 10);
+
+    verify(fileListener, never()).onFileCreate(any(File.class), any(Synchronization.class));
+    verify(fileListener, never()).onFileChange(any(File.class), any(Synchronization.class));
+    verify(fileListener, times(files.length))
+        .onFileDelete(any(File.class), any(Synchronization.class));
+  }
+
+  @Test
+  public void testModifyAndDeleteWithDelays() throws Exception {
+    int delay = 200;
+
+    File[] files = initFiles(1, monitoredDirectory, "file00");
+    observer.checkAndNotify();
+    init();
+
+    Stream.of(files).forEach(this::changeData);
+
+    addDelay(delay);
+
+    observer.checkAndNotify();
+
+    Stream.of(files).forEach(this::fileDelete);
+    //  Implementation Detail
+    //  Because the last operation hasn't finished yet the file should not be deleted in the state.
+    //  It will wait until the file has finished it's prior update to remove it.
+
+    observer.checkAndNotify();
+    verify(fileListener, never()).onFileDelete(any(File.class), any(Synchronization.class));
+    Thread.sleep(delay + delayBuf);
```

Comment:
❓ Is there a chance the onFileChange is not called in time after this sleep finished?

Reply:
That's the reason for the `delayBuf`, I ran these test 50 times before putting up the PR without them failing just to make sure. If it does end up being a problem in the future though, then we can increase the buffer and that will "fix" all the tests

Reply:
Typically we should avoid sleeps in tests since in slower environments these have greater potential to fail.

Reply:
🤔 Would there be another way to efficiently test threads occurring in different intervals?

Reply:
I'm going to take a stab at getting rid of thread sleeps and replace them with synchronized blocks to emulate what thread.sleep would accomplish